### PR TITLE
[KOGITO-5083] - Kogito Runtime Image Stream Fails to Update

### PR DESCRIPTION
**Pull request by vaibhavjainwiz**
Many thanks for submitting your Pull Request :heart:! 

Jira Issue : https://issues.redhat.com/browse/KOGITO-5083

Please make sure your PR meets the following requirements:

- [x] You have read the [contributors' guide](https://github.com/kiegroup/kogito-operator/blob/master/README.md#contributing-to-the-kogito-operator)
- [x] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [x] Pull Request contains a link to the JIRA issue
- [x] Pull Request contains a description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Your feature/bug fix has a unit test that verifies it
- [x] You've tested the new feature/bug fix in an actual OpenShift cluster
- [x] You've added a [RELEASE_NOTES.md](RELEASE_NOTES.md) entry regarding this change

### CODE REVIEW

File: core/infrastructure/imagestream.go [hunk 1]
```diff
@@ -83,32 +102,103 @@ func (i *imageStreamHandler) MustFetchImageStream(key types.NamespacedName) (*im
...
+		tagReference := i.createImageStreamTag(tag, addFromReference, imageName, insecureImageRegistry)
+		tagReferences := append(imageStream.Spec.Tags, tagReference)
+		imageStream.Spec.Tags = tagReferences
+	}
+	return imageStream, nil
+}
+
+// createImageStream creates the ImageStream referencing the given namespace.
+func (i *imageStreamHandler) createImageStream(name, namespace string) *imgv1.ImageStream {
+	i.Log.Debug("Creating new Image stream.")
```
Comment (radtriste) [inline core/infrastructure/imagestream.go:125]: please add the name of the created img stream

File: core/infrastructure/imagestream.go [hunk 2]
```diff
@@ -43,8 +44,10 @@ var imageStreamAnnotations = map[string]string{
 // ImageStreamHandler ...
 type ImageStreamHandler interface {
 	FetchImageStream(key types.NamespacedName) (*imgv1.ImageStream, error)
+	FetchTag(key types.NamespacedName, tag string) (*imgv1.ImageStreamTag, error)
 	MustFetchImageStream(key types.NamespacedName) (*imgv1.ImageStream, error)
-	CreateImageStream(name, namespace, imageName, tag string, addFromReference, insecureImageRegistry bool) *imgv1.ImageStream
+	CreateImageStreamIfNotExists(name, namespace, tag string, addFromReference bool, imageName string, insecureImageRegistry bool) (*imgv1.ImageStream, error)
```
Comment (radtriste) [inline core/infrastructure/imagestream.go:49]: why not using the `types.NamespacedName` in this method ?

**Review by radtriste — APPROVED** [review]
minor comments

**Review by ricardozanini — APPROVED** [review]
Fine by me once @radtriste comments are addressed. :+1:

File: core/infrastructure/imagestream.go (hunk 1 above)
Comment (vaibhavjainwiz) [reply core/infrastructure/imagestream.go]: done

File: core/infrastructure/imagestream.go (hunk 2 above)
Comment (vaibhavjainwiz) [reply core/infrastructure/imagestream.go]: done

**Review by sutaakar — APPROVED** [review]
Looks good to me, fixes the issue described in JIRA.
I have a suggestion for improvement (cam be addressed here or separately). Noticed that if you change the image name (keeping tag the same) then ImageStream is not updated.
For example:
```
apiVersion: app.kiegroup.org/v1beta1
kind: KogitoRuntime
metadata:
  name: example-quarkus
spec:
  # see the quarkus-jvm.Dockerfile in this directory to know how to build your image locally
  image: quay.io/kiegroup/examples-process-quarkus-example-nightly:latest
```
Then ImageStream tag example-quarkus:latest pointing to `quay.io/kiegroup/examples-process-quarkus-example-nightly:latest` is created. However when I edit KogitoRuntime, changing image to `quay.io/kiegroup/examples-process-quarkus-example-nightly2:latest` then ImageStream still points to the old image.

File: core/infrastructure/image.go [hunk 3]
```diff
@@ -84,69 +82,36 @@ func NewImageHandler(context operator.Context, image *api.Image, defaultImageNam
...
 		return imageStream, nil
 	}
 	return nil, nil
 }
 
 // resolveImage resolves images like "quay.io/kiegroup/kogito-jobs-service:latest" or "internal-registry/namespace/image:hash".
 // Can be empty if on OpenShift and the ImageStream is not ready.
+// In case of Openshift, image name is resolved using Image Stream tag to support kogito build.
 func (i *imageHandler) ResolveImage() (string, error) {
+	i.Log.Info("Going to resolve image...")
```
Comment (sutaakar) [inline core/infrastructure/image.go:98]: Can you use debug level log here?

File: core/infrastructure/image.go [hunk 4]
```diff
@@ -84,69 +82,36 @@ func NewImageHandler(context operator.Context, image *api.Image, defaultImageNam
...
 	if i.Client.IsOpenshift() {
-		if err := i.validateImageStreamTagStatus(); err != nil {
+		i.Log.Info("Openshift environment found. Going to resolve image using ImageStream.")
```
Comment (sutaakar) [inline core/infrastructure/image.go:100]: Same as above.

File: core/infrastructure/image.go (hunk 3 above)
Comment (vaibhavjainwiz) [reply core/infrastructure/image.go:98]: done

File: core/infrastructure/image.go (hunk 4 above)
Comment (vaibhavjainwiz) [reply core/infrastructure/image.go:100]: done

**Comment by ricardozanini** [comment]
@sutaakar can you open a JIRA for us to track this?

**Comment by sutaakar** [comment]
Reported as https://issues.redhat.com/browse/KOGITO-5205

**Pull request merged**